src/BucketTesting/Validation/CampaignUtilizationValidator.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\BucketTesting\Validation;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\NodeFinder;
+use PhpParser\ParserFactory;
+use Symfony\Component\Filesystem\Exception\FileNotFoundException;
+use WMDE\Fundraising\Frontend\BucketTesting\Campaign;
+use WMDE\Fundraising\Frontend\BucketTesting\CampaignCollection;
+
+/**
+ * @license GNU GPL v2+
+ */
+class CampaignUtilizationValidator {
+
+	/** @see \WMDE\Fundraising\Frontend\Factories\ChoiceFactory */
+	const CHOICE_FACTORY_LOCATION = 'src/Factories/ChoiceFactory.php';
+	/** @see \WMDE\Fundraising\Frontend\BucketTesting\FeatureToggle::featureIsActive */
+	const FEATURE_TOGGLE_METHOD_NAME = 'featureIsActive';
+
+	private $hasValidated = false;
+	private $campaignCollection;
+	private $errorLogger;
+	private $ignoredBuckets = [];
+
+	public function __construct( CampaignCollection $campaignCollection, array $ignoredBuckets, ValidationErrorLogger $errorLogger ) {
+		$this->campaignCollection = $campaignCollection;
+		$this->errorLogger = $errorLogger;
+		$this->ignoredBuckets = $ignoredBuckets;
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errorLogger->getErrors();
+	}
+
+	private function validate(): void {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		$featureToggleChecks = $this->getFeatureToggleChecks();
+		if ( $duplicates = $this->getDuplicates( $featureToggleChecks ) ) {
+			foreach ( $duplicates as $duplicate ) {
+				$this->errorLogger->addError(
+					'ChoiceFactory contains multiple checks for feature toggle: ' . $duplicate
+				);
+			}
+			return;
+		}
+
+		$configBucketIdList = $this->buildBucketIdsFromCampaignCollection(
+			$this->campaignCollection
+		);
+
+		/** Remove all buckets which were marked as test buckets in TEST_BUCKET_IGNORE_LIST */
+		$featureToggleChecks = array_filter( $featureToggleChecks, [ __CLASS__, 'filterBucketIds' ] );
+		$configBucketIdList = array_filter( $configBucketIdList, [ __CLASS__, 'filterBucketIds' ] );
+
+		/** Check if either the configuration files or the choice factory has rules not found in the other */
+		$codeConfigDiff = array_diff( $featureToggleChecks, $configBucketIdList );
+		$configCodeDiff = array_diff( $configBucketIdList, $featureToggleChecks );
+
+		if ( empty( $codeConfigDiff ) === false || empty( $configCodeDiff ) === false ) {
```

Comment:
The code will become more readable (and less nested) if you invert the check, return early and then do the foreach iterations.